### PR TITLE
IEN-941 | [Undo EoJ] Create Re-Engage milestone

### PR DESCRIPTION
**Pull request by jerry-ey**

### CODE REVIEW

**Comment by jerry-ey** [comment]
Going to fix formatting and sonarqube.

**Comment by jerry-ey** [comment]
Fixed formatting and sonarQube.

File: apps/web/src/utils/milestone-utils.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+export const getNextMilestone = (
+  milestone: ApplicantStatusAuditRO,
+  applicant: ApplicantRO,
+  milestones: ApplicantStatusAuditRO[],
+): ApplicantStatusAuditRO | undefined => {
+  const mergedMilestones =
+    applicant?.jobs
+      ?.map(job => job.status_audit || [])
+      .reduce((a, c) => a.concat(c), [...milestones]) || milestones;
```
Comment (dbayly-freshworks) [inline apps/web/src/utils/milestone-utils.ts:23]: This one liner is a bit complicated. Can you either add a comment explaining the logic here or add more descriptive variable names.

**Review by dbayly-freshworks — APPROVED** [review]
Overall this looks good. 
I have a few nits. One thing that I am somewhat concerned about is large chunks of stying. I don't have a clear answer on the best way to address it, but they are hard to parse, and I think hard to maintain, especially because they are decentralized. Let me know if you have any ideas.

**Comment by jerry-ey** [comment]
All base UIs are from shadcn/ui site, it's maintained there. simply copy-paste. The good thing is it's not like material-ui, you can not change the source code. This shadcn/ui only guides you how to code the ui, so that you are flexible to customize.

For example: https://ui.shadcn.com/docs/components/alert
CleanShot 2024-12-10 at 08 37 15@2x

File: apps/web/src/utils/milestone-utils.ts (hunk 1 above)
Comment (jerry-ey) [reply apps/web/src/utils/milestone-utils.ts:23]: I actually didn't write all the code in this util file, I think it's done by Shahwan, this is a refactor to the file `MilestoneTable.tsx`
Because I copy this table most of the code to my new file `SystemMilestoneTable.tsx` and sonarqube is complaining the duplication, that's why I have to extract out the logic and refactor it. I on purposely not refactor this old file `MilestoneTable.tsx` with this `milestone-utils.ts`, I think I just keep what it was so that PR easier to read.

```tsx
import { findNextMilestone } from '@ien/common/dist/helper/find-next-milestone';
import { Fragment, useEffect, useState } from 'react';

import { AclMask, PageOptions, Pagination } from '@components';
import {
  Access,
  ApplicantStatusAuditRO,
  formatDate,
  IENApplicantUpdateStatusDTO,
  StatusCategory,
} from '@ien/common';
import editIcon from '@assets/img/edit.svg';
import disabledEditIcon from '@assets/img/disabled_edit.svg';
import { getHumanizedDuration, updateMilestone } from '@services';
import { useApplicantContext } from '../applicant/ApplicantContext';
import { useAuthContext } from '../AuthContexts';
import { AddMilestone } from './recruitment/AddMilestone';
import { Milestone } from './recruitment/Milestone';
import dayjs from 'dayjs';

interface MilestoneTableProps {
  category: string | StatusCategory;
}

const DEFAULT_TAB_PAGE_SIZE = 5;
const EDIT_NON_RECRUITMENT_MILESTONES = false;

const getStatus = (milestone: ApplicantStatusAuditRO) => {
  const { status } = milestone.status;
  if (!status) return '';

  const label = milestone.status.party || '-';

  return (
    <div className='flex flex-row'>
      <div className='bg-bcGrayLabel px-2 py-0.5 mr-2 text-xs text-white rounded'>{label}</div>
      <div className='text-ellipsis overflow-hidden ...'>{status}</div>
    </div>
  );
};

const currentlyEditing = 'bg-blue-100';

export const MilestoneTable = ({ category }: MilestoneTableProps) => {
  const { authUser } = useAuthContext();
  const { applicant, milestones, fetchApplicant } = useApplicantContext();

  const [filteredMilestones, setFilteredMilestones] = useState<ApplicantStatusAuditRO[]>([]);
  const [milestonesInPage, setMilestonesInPage] = useState<ApplicantStatusAuditRO[]>([]);
  const [editing, setEditing] = useState<ApplicantStatusAuditRO | null>(null);
  const [activeEdit, setActiveEdit] = useState(0);

  const [pageIndex, setPageIndex] = useState(1);
  const [pageSize, setPageSize] = useState(DEFAULT_TAB_PAGE_SIZE);

  useEffect(
    function filterMilestones() {
      const audits =
        milestones
          ?.filter(audit => {
            return audit.status.category === category;
          })
          .sort((b, a) => {
            if (a.start_date === b.start_date) {
              return dayjs(a.updated_date).diff(b.updated_date);
            }
            return dayjs(a.start_date).diff(b.start_date);
          }) || [];
      setFilteredMilestones(audits);
    },
    [milestones, category, applicant],
  );

  useEffect(
    function updateMilestonesInPage() {
      if (!filteredMilestones || (pageIndex - 1) * pageSize > filteredMilestones.length) {
        setPageIndex(1);
      }
      const start = (pageIndex - 1) * pageSize;
      const end = pageIndex * pageSize;
      setMilestonesInPage(filteredMilestones?.slice(start, end) || []);
    },
    [filteredMilestones, pageIndex, pageSize],
  );

  const handleUpdateMilestone = async (id: string, values: IENApplicantUpdateStatusDTO) => {
    setEditing(null);

    const milestone = await updateMilestone(applicant.id, id, values);
    if (milestone) {
      fetchApplicant();
    }
  };

  const handlePageOptions = (options: PageOptions) => {
    setPageSize(options.pageSize);
    setPageIndex(options.pageIndex);
  };

  const getNextMilestone = (
    milestone: ApplicantStatusAuditRO,
  ): ApplicantStatusAuditRO | undefined => {
    const mergedMilestones =
      applicant?.jobs
        ?.map(job => job.status_audit || [])
        .reduce((a, c) => a.concat(c), [...milestones]) || milestones;
    return findNextMilestone(milestone, mergedMilestones);
  };

  const getDuration = (milestone: ApplicantStatusAuditRO): string => {
    const start = milestone.start_date;
    if (!start) return '-';

    const end = getNextMilestone(milestone)?.start_date;

    if (start === end) return '0 days';

    return getHumanizedDuration(start, end);
  };

  const canAddEditNonRecruitmentMilestone = () => {
    // success: no ats1_id(not from ATS), ha id exists, applicant is part of same ha as logged-in user
    return !applicant.ats1_id && authUser?.ha_pcn_id && !editing;
  };

  return (
    <div>
      <div className='text-bcGray pb-2'>Showing {milestonesInPage.length} logs</div>
      <div>
        <table className='w-full'>
          <thead className=''>
            <tr className='bg-bcLightGray text-bcDeepBlack text-sm text-left border-b-2 border-yellow-300'>
              <th className='py-4 pl-8' scope='col'>
                Milestones
              </th>
              <th className='px-4' scope='col'>
                Date
              </th>
              <th className='px-4' scope='col'>
                Duration
              </th>
              {EDIT_NON_RECRUITMENT_MILESTONES && <th className='px-4' scope='col'></th>}
            </tr>
          </thead>
          <tbody className='text-bcBlack'>
            {milestonesInPage.map((audit, index) => (
              <Fragment key={audit.id}>
                <tr
                  className={`text-left text-sm shadow-xs whitespace-nowrap ${
                    editing && activeEdit === index ? currentlyEditing : 'even:bg-bcLightGray'
                  }`}
                >
                  <td className='pl-8 py-5'>{getStatus(audit)}</td>
                  <td className='px-4'>{formatDate(audit.start_date || '')}</td>
                  <td className='px-4'>{getDuration(audit)}</td>
                  {EDIT_NON_RECRUITMENT_MILESTONES && (
                    <td className='items-center px-4'>
                      <AclMask acl={[Access.APPLICANT_WRITE]}>
                        <button
                          onClick={() => {
                            setEditing(audit);
                            setActiveEdit(index);
                          }}
                          disabled={
                            !canAddEditNonRecruitmentMilestone() || (!!editing && audit === editing)
                          }
                        >
                          <img
                            src={
                              canAddEditNonRecruitmentMilestone()
                                ? editIcon.src
                                : disabledEditIcon.src
                            }
                            alt='edit milestone'
                          />
                        </button>
                      </AclMask>
                    </td>
                  )}
                </tr>
                {EDIT_NON_RECRUITMENT_MILESTONES && (
                  <tr>
                    <td colSpan={5}>
                      <Milestone
                        milestone={audit}
                        editing={editing}
                        onEditing={setEditing}
                        handleSubmit={values => handleUpdateMilestone(audit.id, values)}
                        category={category}
                      />
                    </td>
                  </tr>
                )}
              </Fragment>
            ))}
          </tbody>
        </table>
        {!milestonesInPage.length && (
          <div className='w-full flex flex-row justify-center py-5 font-bold'>No Milestones</div>
        )}
        {/* first check if applicant has write access,
        then only show form if applicant was not added by ATS and if current logged-in user added applicant */}
        <AclMask acl={[Access.APPLICANT_WRITE]}>
          {EDIT_NON_RECRUITMENT_MILESTONES && canAddEditNonRecruitmentMilestone() && (
            <AddMilestone category={category} />
          )}
        </AclMask>
      </div>
      <Pagination
        id='milestone-page'
        pageOptions={{ pageIndex, pageSize, total: filteredMilestones.length }}
        onChange={handlePageOptions}
      />
    </div>
  );
};

```

File: apps/web/package.json [hunk 2]
```diff
@@ -17,25 +17,38 @@
     "@fortawesome/free-solid-svg-icons": "6.5.2",
     "@fortawesome/react-fontawesome": "0.2.2",
     "@headlessui/react": "2.1.2",
+    "@hookform/resolvers": "^3.9.1",
     "@ien/common": "1.0.2",
+    "@radix-ui/react-dialog": "^1.1.2",
+    "@radix-ui/react-label": "^2.1.0",
+    "@radix-ui/react-popover": "^1.1.2",
+    "@radix-ui/react-select": "^2.1.2",
+    "@radix-ui/react-slot": "^1.1.0",
```
[the remark this answers is not in the document]
Comment (jerry-ey) [reply apps/web/package.json:26]: Thank you, I will fix this version things in the next logic PR, IEN-950. so that easier to not have git conflict.

File: apps/web/src/utils/milestone-utils.ts (hunk 1 above)
Comment (jerry-ey) [reply apps/web/src/utils/milestone-utils.ts:23]: I'll add description that this is extract from the file `MilestoneTable.tsx` in next PR.
